core/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingChanges.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.routing.allocation;
+
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.UnassignedInfo;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.index.shard.ShardId;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class RoutingChanges {
+    private final Map<ShardId, ShardRoutingChanges> shardChanges = new HashMap<>();
+
+    private boolean changed;
+
+    /**
+     * Returns whether changes were made
+     */
+    public boolean isChanged() {
+        return changed;
+    }
+
+    public Map<ShardId, ShardRoutingChanges> getChanges() {
+        return Collections.unmodifiableMap(shardChanges);
+    }
+
+    /**
+     * Called when unassigned shard is initialized. Does not include initializing relocation target shards.
+     */
+    public void shardInitialized(ShardRouting unassignedShard) {
+        assert unassignedShard.unassigned() : "expected unassigned shard " + unassignedShard;
+        changes(unassignedShard.shardId()).initializedShards.add(unassignedShard);
+    }
+
+    /**
+     * Called when an initializing shard is started.
+     */
+    public void shardStarted(ShardRouting initializingShard, ShardRouting startedShard) {
+        assert initializingShard.initializing() : "expected initializing shard " + initializingShard;
+        assert startedShard.started() : "expected started shard " + startedShard;
```

Comment:
Out of curiosity, would it be better to change these asserts to be runtime checks instead?